src/library/export/engineprimeexportjob.cpp

```diff
@@ -0,0 +1,586 @@
+    int maxProgress = m_trackRefs.size() + m_crateIds.size() + 2;
+    int currProgress = 0;
+    emit jobMaximum(maxProgress);
+    emit jobProgress(currProgress);
+
+    // Ensure that the database exists, creating an empty one if not.
+    std::unique_ptr<djinterop::database> pDb;
+    try {
+        bool created;
+        pDb = std::make_unique<djinterop::database>(el::create_or_load_database(
+                m_request.engineLibraryDbDir.path().toStdString(),
+                m_request.exportVersion,
+                created));
+    } catch (std::exception& e) {
+        qCritical() << "Failed to create/load database:" << e.what();
+        emit failed(e.what());
+        return;
+    }
+
+    ++currProgress;
+    emit jobProgress(currProgress);
+
+    // We will build up a map from Mixxx track id to EL track id during export.
+    QHash<TrackId, int64_t> mixxxToEnginePrimeTrackIdMap;
+
+    for (const auto& trackRef : m_trackRefs) {
+        // Load each track.
+        // Note that loading must happen on the same thread as the track collection
+        // manager, which is not the same as this method's worker thread.
+        QMetaObject::invokeMethod(
+                this,
+                "loadTrack",
+                Qt::BlockingQueuedConnection,
+                Q_ARG(TrackRef, trackRef));
+
+        if (m_cancellationRequested.loadAcquire() != 0) {
+            qInfo() << "Cancelling export";
+            return;
+        }
+
+        DEBUG_ASSERT(m_pLastLoadedTrack != nullptr);
+
+        qInfo() << "Exporting track" << m_pLastLoadedTrack->getId().value()
+                << "at" << m_pLastLoadedTrack->getFileInfo().location() << "...";
+        try {
+            exportTrack(m_request,
+                    pDb.get(),
+                    &mixxxToEnginePrimeTrackIdMap,
+                    m_pLastLoadedTrack,
+                    std::move(m_pLastLoadedWaveform));
+        } catch (std::exception& e) {
+            qCritical() << "Failed to export track"
+                        << m_pLastLoadedTrack->getId().value() << ":"
+                        << e.what();
+            emit failed(e.what());
+            return;
+        }
+
+        m_pLastLoadedTrack.reset();
+
+        ++currProgress;
+        emit jobProgress(currProgress);
+    }
+
+    // We will ensure that there is a special top-level crate representing the
+    // root of all Mixxx-exported items.  Mixxx tracks and crates will exist
+    // underneath this crate.
+    std::unique_ptr<djinterop::crate> pExtRootCrate;
+    try {
+        const auto optionalExtRootCrate = pDb->root_crate_by_name(kMixxxRootCrateName);
+        pExtRootCrate = std::make_unique<djinterop::crate>(optionalExtRootCrate
+                        ? *optionalExtRootCrate
+                        : pDb->create_root_crate(kMixxxRootCrateName));
+    } catch (std::exception& e) {
+        qCritical() << "Failed to create/identify root crate:" << e.what();
```

Comment:
Does this exception leave Mixxx in an inconsistent state or causes undefined behavior? If not I would lower the *log level* to warning. This reasoning should be applied to all qCritical() logs.

And qFatal() should only be used if the application is about to crash.

Reply:
My reasoning for `qCritical()` was that this error is critical to the _export process_, but of course it's not fatal to Mixxx overall.  I used it as I might use an 'ERROR' logging level in other logging frameworks.  I don't believe I've used `qFatal()` anywhere in this PR.

Happy to take guidance here on what the convention is within Mixxx.

Reply:
The export process might be incomplete. As long as the resulting database is consistent and usable I would not consider this a hard error that requires *immediate human intervention*. See also: https://stackoverflow.com/a/8021604

Ideally the user should be informed. But since we don't have the *notification center* yet this is difficult to achieve. You may leave a TODO that explains the consequences and possible options. This is currently missing. The log only mentions the failed operation, but not how to deal with it.

qFatal() was just an example. I only wanted to point out that it is even more serious than qCritical() (= error).

Reply:
So far, if _any_ of these exceptions occur, then:

* Mixxx and its database should never be left in an inconsistent state.  The export is pretty much read-only from the Mixxx perspective.
* The exported Engine Prime DB should never be left in an inconsistent state either.  It may be the case that a given track/crate doesn't get exported, but the existing DB remains usable.  This is a conscious design choice in the libdjinterop API.

Right now, if something fails in the export process, the code logs at `qCritical()` and also emits the `failed` signal, which is connected to an error dialog (a call to `QMessageBox::critical()` in `libraryexporter.cpp`) that will inform the user.  A message box like that could be replaced in future if/when notification centre-like functionality is introduced as you say.

I could add a TODO to the message box code in `libraryexporter.cpp`, mentioning the possibility of a better means of notification - no issues with this.

With regard to using _immediate human intervention_ as a criterion for whether something is a critical error or not, I think this might be a subjective topic.  The human here is the person who wishes to export a sub-set of their Mixxx library to Engine Prime, presumably to take to a gig on a USB stick or similar.  One could make the argument that if the export failed then they can't do that and their immediate intervention is indeed required!

@uklotzde I don't have a strong opinion here, so if you'd like me to change anything about this aspect to better align with Mixxx conventions then I'm happy to do so - just let me know what you prefer.  I'll add the TODO about better notifications anyway.

Reply:
In my opinion qCritical() should be reserved for unresolvable inconsistencies and data corruption, unexpected failures of the OS, or hardware error.

If you search for qCritical() in out code base you will only find very few occurrences that cause a real malfunction of Mixxx. We should not stress this terminology too much. I also recommend to reword the critical() signal for consistency. Either failed() or errorOccurred() might be more appropriate.

Example: The networking tasks use failed() for application-level protocol errors and networkError() for low-level transport errors. Non of those issues is logged as critical.

Reply:
Ok - I've pushed a commit that logs export failures as `qWarning()`, and there is now a TODO about presenting the results of an export to the user in a better if notification functionality is added in future.

The failure signal is called `failed` (it was previously too).